### ISSUE TITLE
"Requested transaction contradicts existing jobs" messages on shutdown

### ISSUE THREAD

**Issue by c4rlo** [Vladimir-csp/uwsm#30]
Since uwsm 0.16.0 (and also with 0.16.1), I've noticed some odd trace in the systemd journal on shutdown. It happens with `reboot` and `systemctl soft-reboot`, but not with `loginctl terminate-user ''`.

```
Sun 2024-06-30 11:18:25 BST uwsm_sway[1434]: 00:01:46.471 [ERROR] [wlr] [libseat] [libseat/backend/logind.c:199] Could not close device: Remote peer disconnected
Sun 2024-06-30 11:18:25 BST uwsm_sway[1434]: 00:01:46.472 [ERROR] [wlr] [backend/session/session.c:356] Failed to close device 36: Resource temporarily unavailable
Sun 2024-06-30 11:18:25 BST systemd[1]: Requested transaction contradicts existing jobs: Transaction for systemd-logind.service/start is destructive (user.slice has 'stop' job queued, but 'start' is included in transaction).
Sun 2024-06-30 11:18:25 BST systemd[1]: Requested transaction contradicts existing jobs: Transaction for systemd-logind.service/start is destructive (user.slice has 'stop' job queued, but 'start' is included in transaction).
Sun 2024-06-30 11:18:25 BST uwsm_sway[1434]: 00:01:46.499 [ERROR] [wlr] [libseat] [libseat/backend/logind.c:199] Could not close device: Could not activate remote peer 'org.freedesktop.login1': activation request failed: a concurrent deactivation request is already in progress
Sun 2024-06-30 11:18:25 BST uwsm_sway[1434]: 00:01:46.499 [ERROR] [wlr] [backend/session/session.c:356] Failed to close device 37: Resource temporarily unavailable
Sun 2024-06-30 11:18:25 BST systemd[1]: Requested transaction contradicts existing jobs: Transaction for systemd-logind.service/start is destructive (veritysetup.target has 'stop' job queued, but 'start' is included in transaction).
[...several repetitions of similar trace...]
Sun 2024-06-30 11:18:25 BST uwsm_sway[1434]: 00:01:47.298 [ERROR] [wlr] [backend/session/session.c:356] Failed to close device 12: Resource temporarily unavailable
Sun 2024-06-30 11:18:25 BST systemd[1]: Requested transaction contradicts existing jobs: Transaction for systemd-logind.service/start is destructive (boot.mount has 'stop' job queued, but 'start' is included in transaction).
Sun 2024-06-30 11:18:25 BST uwsm_sway[1434]: 00:01:47.298 [ERROR] [wlr] [libseat] [libseat/backend/logind.c:365] Could not release control of session: Could not activate remote peer 'org.freedesktop.login1': activation request failed: a concurrent deactivation request is already in progress
Sun 2024-06-30 11:18:26 BST systemd[1350]: Stopped Main service for sway.
Sun 2024-06-30 11:18:26 BST systemd[1350]: wayland-wm@sway.service: Triggering OnSuccess= dependencies.
Sun 2024-06-30 11:18:26 BST systemd[1350]: Stopped target Session services which should run early before the graphical session is brought up.
Sun 2024-06-30 11:18:26 BST systemd[1350]: Stopped target Preparation for session of sway Wayland compositor.
Sun 2024-06-30 11:18:26 BST systemd[1350]: Reached target Shutdown graphical session units.
Sun 2024-06-30 11:18:26 BST systemd[1350]: Stopping Environment preloader for sway...
Sun 2024-06-30 11:18:26 BST systemd[1350]: Stopped target Shutdown graphical session units.
```

It seems to be harmless, so maybe it's a non-issue, but I wanted to mention it. I'm using systemd 256.1, sway 1.9, Arch Linux.

It also feels like shutdown is a tiny bit slower than with uwsm 0.15.8. I'll experiment a bit more to see what is going on.

**Comment by c4rlo** [comment]
I have confirmed that this definitely doesn't happen with uwsm 0.15.8.

**Comment by Vladimir-csp** [comment]
This is strange. User-level units should not affect system-level units via dependencies in any way.

I've found `systemd-logind.service/start is destructive` messages on shutdown all the way back to October, when neither `wayland-session-shutdown.target` nor `wayland-session-bindpid@.service` did exist. This seems to be happening on my machine for a long time.

Found one such message from a user manager though, during a failed wayland session startup attempt. I will add `replace-irreversibly` to transactions invoking `wayland-session-shutdown.target` analogous to what systemd does to `shutdown.target` on system level. But this should just remove some errors from user manager in corner cases and would not affect system level.

If you have any ideas on why it might depend on uwsm versions on your system, but not on mine, it would be very helpful.

**Comment by c4rlo** [comment]
Thanks for looking into it!

I've also looked a bit more into it, first by increasing the systemd log verbosity, then by using `busctl monitor` to log all D-Bus messages during that time (via a little systemd service), but neither of these have really given me any more clues; I couldn't figure out why systemd is getting a job to start `systemd-logind` at that time.

Looking a bit more closely at the trace though, it seems that what must be going on is that sway wants to tell logind to close the session, and since logind is currently shutting down, this triggers some kind of socket activation where systemd wants to start up logind, hence it then complains that it is in the process of shutting down.

So, maybe one of the uwsm-managed services (e.g. `wayland-wm@.service`) needs to have an `After=` on either `systemd-logind.service`, or maybe on the unit that contains the `login` service (which for me is currently `session-43.scope`)...?

**Comment by c4rlo** [comment]
Indeed, focusing on this part of the trace, which I see repeated numerously:

```
Sun 2024-06-30 11:18:25 BST uwsm_sway[1434]: 00:01:46.499 [ERROR] [wlr] [libseat] [libseat/backend/logind.c:199] Could not close device: Could not activate remote peer 'org.freedesktop.login1': activation request failed: a concurrent deactivation request is already in progress
```

The corresponding code is [here](https://git.sr.ht/~kennylevinsen/seatd/tree/0.8.0/item/libseat/backend/logind.c#L199) in libseat, which gets called by wlroots [here](https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/backend/session/session.c#L352), and ultimately will be part of sway shutdown.

**Comment by Vladimir-csp** [comment]
> sway wants to tell logind to close the session

This is actually funny.

> So, maybe one of the uwsm-managed services (e.g. wayland-wm@.service) needs to have an After= on either systemd-logind.service, or maybe on the unit that contains the login service (which for me is currently session-43.scope)...?

User-level and system-level units can not depend on each other. The closest system unit is `user@.service` which is the template user's instances of systemd, which contain user-level units. 

I've added `After=systemd-logind.service` to  `user@.service`, destructive transactions on soft-reboot seem to be gone.  But I'm not sure how good of a fix this is.

**Comment by Vladimir-csp** [comment]
It seems to be a race condition. Since there is no ordering between `user@.service` and `systemd-logind.service`, slower shutdown of the former causes this situation. It is slower because uwsm it makes the compositor part of the ordered structure of graphical-session-pre, graphical-session, xdg-autostart. So compositor's shutdown is delayed until graphical autosart apps are brought down, then other graphical apps are brought down. Meanwhile, systemd-logind is stopped.

At least that is my hypothesis about your system. When I tried to test this on my system by launching a simple service
```
[Service]
Type=exec
ExecStart=sway -c /dev/null
```
with nothing else around and doing a soft-reboot, logind still left the picture before sway.

**Comment by Vladimir-csp** [comment]
Good news, this is [being resolved](https://github.com/systemd/systemd/pull/33555) upstream.

**Comment by c4rlo** [comment]
Yeah, fair enough. Not something for uwsm to solve (and also a very minor issue to begin with). Feel free to close this again, I had just wanted to give an update.

Thanks for your help again, I learnt some good things here :)

**Comment by Vladimir-csp** [comment]
> and changing the final `systemctl --user start --wait wayland-session-bindpid` to run with `SIGTERM` being `SIG_IGN`ored. 

There is something wrong here, bindpid unit is not started with `--wait`, did you mean `wayland-wm@${compositor}.service`?

Stupid question: how do you run a process shielded from a signal?

**Comment by c4rlo** [comment]
> > and changing the final `systemctl --user start --wait wayland-session-bindpid` to run with `SIGTERM` being `SIG_IGN`ored.
> 
> There is something wrong here, bindpid unit is not started with `--wait`, did you mean `wayland-wm@${compositor}.service`?

Yes, you're right. I also did that experiment a bit hastily, not sure how much sense it made any more, I'll take a deeper look later when I have more time.

> Stupid question: how do you run a process shielded from a signal?

By doing `signal.signal(signal.SIGTERM, signal.SIG_IGN)` before the `execlp()`.

**Comment by Vladimir-csp** [comment]
Shielding a waiting `systemctl` is counterproductive. It needs to be `TERM`able in order for bindpid unit to work.

To hold session open, there should be another process that either ignores `TERM` and `HUP` or uses them as a trigger to check on units state, then exit only after they are deactivated.

**Comment by c4rlo** [comment]
I got it to work now by doing approximately what uwsm 0.15.8 did:

```diff
diff --git a/uwsm/main.py b/uwsm/main.py
index 8be8975..83024b9 100644
--- a/uwsm/main.py
+++ b/uwsm/main.py
@@ -3859,6 +3859,20 @@ def stop_wm():
     return True
 
 
+exiting = False
+
+def on_exit_signal(signum, frame):
+    global exiting
+    if exiting:
+        print_normal(f"Got signal {signum}: already exiting so ignoring")
+        return
+    exiting = True
+    print_normal(f"Got signal {signum}: stopping compositor then exiting")
+    stop_wm()
+    print_normal("Stopped compositor, exiting")
+    sys.exit()
+
+
 def main():
     "UWSM main entrypoint"
 
@@ -3982,17 +3996,23 @@ def main():
             )
             print_debug(sprc)
 
-            # replace ourselves with systemctl
+            signal.signal(signal.SIGTERM, on_exit_signal)
+            signal.signal(signal.SIGINT, on_exit_signal)
+            signal.signal(signal.SIGHUP, signal.SIG_IGN)
+
             # this will start the main compositor unit
             # and wait until compositor is stopped
-            os.execlp(
-                "systemctl",
-                "systemctl",
-                "--user",
-                "start",
-                "--wait",
-                f"wayland-wm@{CompGlobals.id_unit_string}.service",
+            sprc = subprocess.run(
+                [
+                    "systemctl",
+                    "--user",
+                    "start",
+                    "--wait",
+                    f"wayland-wm@{CompGlobals.id_unit_string}.service",
+                ],
+                check=True,
             )
+            print_debug(sprc)
 
         except Exception as caught_exception:
             print_error(caught_exception)
```

This gets rid of the offending trace when sway exits. I guess now I could also remove `wayland-session-bindpid@.service` entirely again, effectively reverting 6d6bdaf, but I haven't tried that. I also suspect you had a good reason for that commit – curious what it was?

Btw, for some reason my change introduces a shutdown delay of about 4 seconds in some random place:

```
Mon 2024-07-01 21:00:10 BST systemd[3480]: Stopping Display colour temperature adjustment...
Mon 2024-07-01 21:00:10 BST thermald[3441]: Terminating ...
Mon 2024-07-01 21:00:10 BST systemd[3480]: Stopping Idle manager for Wayland...
Mon 2024-07-01 21:00:10 BST systemd[1]: Stopping Thermal Daemon Service...
Mon 2024-07-01 21:00:10 BST systemd[1]: Stopping Daemon for power management...
[…]
Mon 2024-07-01 21:00:10 BST systemd[1]: Stopped Daemon for power management.
Mon 2024-07-01 21:00:11 BST thermald[3441]: terminating on user request ..
Mon 2024-07-01 21:00:12 BST systemd[1]: thermald.service: Deactivated successfully.
Mon 2024-07-01 21:00:12 BST systemd[1]: Stopped Thermal Daemon Service.
Mon 2024-07-01 21:00:14 BST systemd[3480]: Stopped Display colour temperature adjustment.
```

Though that is almost certainly a separate issue and maybe beside the point here.

**Comment by Vladimir-csp** [comment]
> This gets rid of the offending trace when sway exits. I guess now I could also remove wayland-session-bindpid@.service entirely again, effectively reverting https://github.com/Vladimir-csp/uwsm/commit/6d6bdafe8b6da418226f048df62449269ec560c4, but I haven't tried that. I also suspect you had a good reason for that commit – curious what it was?

I did some experimenting with a couple of display managers, some of them strip session of resources rather abruptly. `uwsm` process crashed because it could not even print anything. Maybe it is their bug, but since `uwsm` relies heavily on systemd mechanisms anyway, I decided to get rid of running `uwsm` process entirely. It now sets the stage, pushes things into motion and steps aside. All the logic and killing is done by systemd (and waitpid).

**Comment by c4rlo** [comment]
I can confirm that 0.16.3 fixes this for me – thanks a lot!
